### PR TITLE
feat(metrics): add equalized_odds fairness metric with test(closes #17)

### LINKED ISSUE #17: Add Equalized Odds for subgroup fairness analysis
### 🎯 Objective
Add Equalized Odds to better evaluate fairness across subgroups by measuring differences in True Positive Rate (TPR) and False Positive Rate (FPR).

---

### 🧐 Why This Matters

Accuracy alone can hide important fairness issues.

A model might have similar accuracy across groups but still behave very differently:
- one group gets more false negatives (low TPR)
- another gets more false positives (high FPR)

Equalized Odds helps surface these differences.

---

### 🛠️ Task

Implement:

```python
equalized_odds(y_true, y_pred, sensitive_features) -> dict
```

in:
```python
trustlens/metrics/bias.py
```

### 📦 Expected Output
```python
result = equalized_odds(y_true, y_pred, {"gender": gender_arr})

# {
#   "gender": {
#     "M": {"tpr": 0.88, "fpr": 0.12},
#     "F": {"tpr": 0.61, "fpr": 0.08},
#     "__summary__": {
#       "tpr_gap": 0.27,
#       "fpr_gap": 0.04,
#       "violation": "severe"
#     }
#   }
# }
```

### ⚖️ Violation Thresholds
`> 0.15` → "severe"
`0.05–0.15` → "moderate"
`< 0.05` → "acceptable"

### 💡 Implementation Hints
- Use sklearn.metrics:
    - recall_score for TPR
    - confusion_matrix or manual computation for FPR

Follow existing patterns in:
```python
trustlens/metrics/bias.py
```
- Ensure output is JSON-serializable
- Handle edge cases (e.g., division by zero)

### 📚 Reference
Hardt, Price, Srebro (2016)
Equality of Opportunity in Supervised Learning (NeurIPS)

### ✅ Acceptance Criteria
 - [x] Function implemented with correct API
 - [x] Per-group TPR and FPR computed
 - [x] Summary metrics (gaps + violation level) included
 - [x] Unit tests added (test_bias.py)
 - [x] Edge cases handled
 - [x] Documentation updated

### 🙌 Contribution

If you’d like to work on this, please comment before starting so we can coordinate.
Happy to guide implementation if needed 🚀

### PR DESCRIPTION
**Pull request by komoike-oss28-ui**
## Summary
Implements `equalized_odds()` fairness metric in `trustlens/metrics/bias.py`.
Returns per-group TPR/FPR with violation level classification (`acceptable` / `moderate` / `severe`).

## Related Issue
Closes #17

## Type of Change
- [x] 🚀 New feature (non-breaking change which adds functionality)

## Changes Made
- [x] Added `equalized_odds()` to `trustlens/metrics/bias.py` following the same structure as `subgroup_performance()`
- [x] Added `_violation_level()` helper function for gap severity classification
- [x] Added `TestEqualizedOdds` test class to `tests/test_bias.py` (6 test cases)
- [x] Updated `CHANGELOG.md` under `[Unreleased]`

## Testing
- [x] All unit tests pass locally.
- [x] Added new tests for the changes made.

## Pre-submit Checklist
- [x] Run `pre-commit run --all-files` and all hooks passed.
- [x] Verified that all tests pass (`make test`).
- [x] Updated the `CHANGELOG.md` if applicable.
- [x] Verified that documentation has been updated for new features.

## Notes for Reviewers
Output structure example:
{"gender": {"0": {"n_samples": 4, "tpr": 0.5, "fpr": 0.0}, "1": {"n_samples": 4, "tpr": 1.0, "fpr": 0.5}, "__summary__": {"tpr_gap": 0.5, "fpr_gap": 0.5, "tpr_violation": "severe", "fpr_violation": "severe"}}}

### CODE REVIEW

**Comment by komoike-oss28-ui** [comment]
Sorry for missing the checkboxes in the template! All items have been completed:

- ✅ New feature (`equalized_odds()` fairness metric)
- ✅ All unit tests pass locally
- ✅ Added new tests (`TestEqualizedOdds` — 6 test cases)
- ✅ `pre-commit run --all-files` passed (trailing whitespace, ruff, ruff format)
- ✅ `make test` passed
- ✅ `CHANGELOG.md` updated under `[Unreleased]`

**Review by Khanz9664 — CHANGES_REQUESTED** [review]
Hey @komoike-oss28-ui — this is a fantastic contribution, especially for a first PR 🙌

You’ve done a really solid job here:

* clean implementation
* strong test coverage (including edge cases + JSON serialization 👍)
* consistent use of sklearn
* clear and well-structured output format

This is exactly the kind of feature TrustLens needs 

---

###  Minor Suggestions Before Merge

Just a few small refinements to align with library design:

---

**1. Avoid rounding inside core metric**

Currently:

```python
"tpr": round(tpr, 4),
"fpr": round(fpr, 4),
```

It’s better to return raw floats and handle rounding at the report/display layer:

```python
"tpr": float(tpr),
"fpr": float(fpr),
```

---

**2. Simplify confusion_matrix guard**

This condition:

```python
if len(np.unique(y_true_g)) >= 1:
```

is always true for non-empty arrays.

You can simplify/remove it safely, or just ensure:

```python
if len(y_true_g) > 0:
```

---

**3. CHANGELOG formatting**

Small formatting tweak — please place the entry under:

```markdown
## [Unreleased]

### Added
- equalized_odds() fairness metric with per-group TPR/FPR and violation analysis
```

---

**4. PR Description Update**

Could you please update the PR description to include:

* a short summary of what was implemented
* a brief example of the output structure
* confirmation of tests added

This helps future contributors understand the change more easily 👍

---

### Design Note

Using separate fields:

```text
tpr_violation / fpr_violation
```

is a great design choice — much clearer than a single combined violation field. Let’s keep this 👌

---

Once these small changes are in, this should be good to merge.

Great work again — really appreciate the quality here

**Comment by komoike-oss28-ui** [comment]
Thank you for the detailed review and kind words! 

I've addressed all the feedback:

1. ✅ Removed `round()` from `tpr`/`fpr` — now returning raw `float` values
2. ✅ Simplified the confusion matrix guard to `if len(y_true_g) > 0`
3. ✅ Fixed `CHANGELOG.md` formatting with proper `### Added` section
4. ✅ Updated PR description with summary, output structure example, and test confirmation

Please let me know if anything else needs adjusting!

**Comment by q404365631** [comment]
Great implementation of the equalized_odds fairness metric! The code follows the established patterns well and includes comprehensive tests. A few suggestions for improvement:

1. Consider adding docstring examples showing the expected output format
2. You might want to add input validation for edge cases (empty groups, etc.)
3. Consider adding a parameter to control violation threshold sensitivity

Overall looks solid - ready for merge once these minor improvements are addressed.

**Comment by q404365631** [comment]
Excellent work on the equalized_odds fairness metric implementation! The code quality is very good and the tests are comprehensive.

A couple of quick observations:
1. The violation level classification logic looks solid - great use of the severity thresholds
2. Consider adding input validation for edge cases like empty groups or single-class scenarios
3. The docstring could benefit from a concrete example showing the expected output format

Overall this is a well-implemented feature that follows the established patterns in the codebase. Looking forward to seeing it merged!

**Comment by komoike-oss28-ui** [comment]
Thank you both for the feedback!

These are great suggestions. I'll defer to @Khanz9664 on whether to address them in this PR or as follow-up issues, to keep the scope focused on the original #17 requirements.

**Review by Khanz9664 — APPROVED** [review]
Thanks @komoike-oss28-ui — this looks great now 👍

All requested changes have been addressed cleanly, and the implementation aligns well with the original scope of #17.

---

###  On Additional Suggestions

Thanks @q404365631 for the thoughtful suggestions — these are definitely valuable improvements.

To keep this PR focused and avoid scope creep, we’ll track these as follow-up enhancements:

* docstring example improvements
* additional input validation
* configurable violation thresholds

These can be handled in separate issues/PRs so we keep this contribution scoped and easy to review 👍

---

###  Decision

This PR fully satisfies the requirements of #17 and is ready to merge.

Great work @komoike-oss28-ui — especially the test coverage and attention to edge cases 🔥

**Comment by Khanz9664** [comment]
Hey @komoike-oss28-ui — quick note 👍

I noticed the commits are showing a different name (`fuchan12040502-innu`). This usually happens due to local Git configuration.

If you want your contributions to be correctly attributed on GitHub, you may want to ensure your Git email is linked to your GitHub account.

Not required for this PR at all — just a small tip for future contributions 🙂

Great work again on this feature 🚀

**Comment by Khanz9664** [comment]
Follow-up issue created for suggested improvements:

👉 #41  

Thanks again for the suggestions and implementation — feel free to pick it up if interested 👍

**Comment by komoike-oss28-ui** [comment]
Thank you so much @Khanz9664! Really glad the implementation met the requirements 🙏

Also appreciate the heads-up on the Git configuration — I'll get that sorted for future contributions.

Looking forward to contributing more to TrustLens! 🚀

**Comment by Khanz9664** [comment]
Really appreciate the contribution @komoike-oss28-ui — this was a great addition to the project 🙌

The implementation quality and test coverage were excellent, especially for a first PR 👏

If you're interested, we’ve opened a follow-up issue (#41) based on some additional ideas discussed in the review. Feel free to pick it up if it looks interesting — no pressure at all 👍

Looking forward to more contributions from you.
